plugins/uefi-capsule/fu-uefi-bootmgr.c

```diff
@@ -249,6 +249,147 @@ fu_uefi_setup_bootnext_with_loadopt(FuEfiLoadOption *loadopt, GError **error)
 	return TRUE;
 }
 
+static GHashTable *
+fu_uefi_bootmgr_parse_csv(const gchar *data, size_t size, GError **error)
+{
+	g_auto(GStrv) lines = NULL;
+	g_autoptr(GHashTable) hash = NULL;
+	g_autofree gchar *nt_data = NULL;
+
+	hash = g_hash_table_new_full(g_str_hash, g_str_equal, g_free, g_free);
+
+	/* null terminate the file */
+	nt_data = g_strndup(data, size);
+
+	lines = g_strsplit(nt_data, "\n", -1);
+
+	for (guint i = 0; lines[i] != NULL; i++) {
+		gchar *endptr = NULL;
+		g_auto(GStrv) columns = NULL;
+		gint64 version;
+		g_autofree gint64 *version_bucket = NULL;
+
+		if ((lines[i][0] == '\0') && (lines[i + 1] == NULL)) {
+			/* last line may be empty */
+			break;
+		}
+
+		columns = g_strsplit(lines[i], ",", 3);
+		if (g_strv_length(columns) < 2) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "invalid CSV data: not enough columns");
+			return NULL;
+		}
+
+		version = g_ascii_strtoll(columns[1], &endptr, 10);
```

Comment:
Can you use fu_strtoull here?

Reply:
hang on-- the version can be a semver no? e.g. fwupd has `1.9.2` -- i think you should store the string rather than try an parse as a uint -- and then use fu_version_compare() to compare them.

Reply:
Actually I should call it "generation" rather than version. That is the correct name. It cannot be a semantic version. It is a number.